packages/optimizer/src/stylable-optimizer.ts

```diff
@@ -49,16 +67,126 @@ export class StylableOptimizer implements IStylableOptimizer {
+                    if (classNamespaceOptimizations) {
+                        node.name = this.getClassName(node.name);
+                    } else if (shortNamespaces) {
+                        const namespaceMatch = node.name.match(namespaceRegexp);
+                        if (!namespaceMatch) {
+                            throw new Error(
+                                `Stylable class dose not have proper namespace ${node.name}`
+                            );
+                        }
+                        node.name = node.name.replace(
+                            namespaceRegexp,
+                            `${this.getNamespace(namespaceMatch[1])}${delimiter}`
+                        );
+                    }
+                }
+            }
+        });
+        return stringifySelector(ast);
+    }
+
+    public optimizeAstAndExports(
+        ast: Root,
+        exported: Record<string, string>,
+        classes = Object.keys(exported),
+        usageMapping: Record<string, boolean>,
+        globals: Record<string, boolean> = {},
+        delimiter?: string,
+        shortNamespaces?: boolean,
+        classNamespaceOptimizations?: boolean
+    ) {
+        if (!shortNamespaces && !classNamespaceOptimizations) {
+            return;
+        }
+        if (!delimiter) {
+            throw new Error(
+                'Missing delimiter when shortNamespaces or classNamespaceOptimizations is enabled'
             );
         }
+
+        ast.walkRules((rule) => {
+            rule.selector = this.rewriteSelector(
+                rule.selector,
+                usageMapping,
+                globals,
+                shortNamespaces || false,
+                classNamespaceOptimizations || false,
+                delimiter
+            );
+        });
+        const namespaceRegexp = new RegExp(`^(.*?)${delimiter}`);
+
+        classes.forEach((originName) => {
+            if (exported[originName]) {
+                exported[originName] = exported[originName]
+                    .split(' ')
+                    .map((renderedNamed) => {
+                        if (classNamespaceOptimizations) {
+                            return this.getClassName(renderedNamed);
+                        } else if (shortNamespaces) {
+                            const namespaceMatch = renderedNamed.match(namespaceRegexp);
+                            if (!namespaceMatch) {
+                                throw new Error(
+                                    `Stylable class dose not have proper export namespace ${renderedNamed}`
+                                );
+                            }
+                            return renderedNamed.replace(
+                                namespaceRegexp,
+                                `${this.getNamespace(namespaceMatch[1])}${delimiter}`
+                            );
+                        } else {
+                            throw new Error('Invalid optimization config');
```

Comment:
better error message (what's invalid about the config)? validate options in the start of optimization and not mid-process?